### PR TITLE
TETRAEDGE: New engine for macOS versions of Syberia 1/2 and Amerzone

### PR DESCRIPTION
**Pull request by mduggan**
### Summary

A new engine for the 2.5D Syberia 1 and 2 and Amerzone macOS editions. Currently only supports Syberia 1 from GOG.  Game can be completed, with some known glitches (below).

### Must TODO Before Merge

~I think this is ready for review, but still has one large change to make before it can be merged - will leave WIP until done to avoid merge:~

- ~Extract OpenGL-specific code to allow TingGL usage (currently doesn't build on platforms without system GL header)~ **DONE**.

Other known issues are small enough that I think they can be worked on in-tree.

### Known bugs

In rough order of severity/annoyance:

- Can sometimes walk during dialogs by clicking around (if done at exactly the wrong time can soft-lock the game) (Partially fixed, still possible)
- ~Walking always does one loop even when character is right next to the object~ Fixed.
- Continue button on main menu should load last save game (currently always disabled)
- Lighting seems wrong - characters go from full light to full shadow, sometimes look a bit too green, etc
- Fuzzy grey artefacts appear around scene objects when they're alpha blended - they go toward black instead of the background, same happens to the "fades" on the splash screens
- Line artefacts sometimes appear at the bottom of text areas
- Bonus menu not fully implemented, and obtained bonuses are not saved in settings data
- Scrolling dialog (for longer text) doesn't get clipped to the dialog area as it scrolls up (have to find how this works in original)
- Footstep sounds are set but not played - still haven't discovered where these are triggered in the original
- The cutscenes that should be 16:9 get stretched to 4:3
- Credits background scrolls too fast, revealing inventory button

### Other notes

- There is a `MATH` commit in here to add a simple Triangle Intersection check to Ray, I'm happy to pull that out into a separate change if preferred
- There is some overlap with the `Te` classes and things in `math/`.  It would be good to merge them some time in the future - at first I reproduced the original closely to avoid bugs from different behaviour there
- Currently only Syberia 1 is implemented - it will need refactoring to support other games, but generally all the code in `te/` should be reusable and all the code in `game/` may need adaption for other games
- From initial investigation it looks like Syberia 2 is a moderately easy addition, but Amerzone uses a quite different version of the engine so will need considerable work

### CODE REVIEW

**Review by aquadran — CHANGES_REQUESTED** [review]
please switch to multi renderer model like in our engines.
and current GL code as "classic GL" with up to 1.4 GL, best keep low as possible

**Comment by bluegr** [comment]
Nice work!

Amerzone is using a different engine, called PhoenixVR, which uses VR scenes and no real-time 3D graphics. Here are some games that used PhoenixVR, with the most notable ones being Amerzone, Dracula 1 + 2 and The Mystery of the Mummy:

https://www.mobygames.com/game-group/middleware-phoenix-vr

If PhoenixVR is using the same file formats as Mac Syberia 1 + 2, that'd be awesome, but I believe it's a totally different engine

**Comment by mduggan** [comment]
> Nice work!
> 
> Amerzone is using a different engine, called PhoenixVR, which uses VR scenes and no real-time 3D graphics. Here are some games that used PhoenixVR, with the most notable ones being Amerzone, Dracula 1 + 2 and The Mystery of the Mummy:
> 
> https://www.mobygames.com/game-group/middleware-phoenix-vr
> 
> If PhoenixVR is using the same file formats as Mac Syberia 1 + 2, that'd be awesome, but I believe it's a totally different engine

I should have been more narrow - I agree for the Windows version of Amerzone.  Similarly the Windows version of Syberia 1/2 use Virtools, so are totally different.

But, the macOS versions of all of them are Tetraedge - eg the GOG version of Amerzone I have installed on my mac, it's using Tetraedge engine. The GUI definitions all use the same format. Eg: 
```
/Applications/GOG/Amerzone The Explorer's Legacy.app/Contents/Resources/game/Amerzone.app/Contents/Resources/GUI> cat answer.lua
TeButtonLayout {
	name = "answer",
	validationSound = "Sounds/SFX/ChoixMenu.ogg",
	upLayout = TeSpriteLayout {
		image = "2D/Answer.png",
	},
	downLayout = TeSpriteLayout {
		image = "2D/AnswerDown.png",
	},
	hitZone = TeLayout {
		sizeType = Percent, size = {1.0, 1.0},
	},
	TeTextLayout {
		name = "text",
		anchor = {0.0, 0.5},
		position = {0.0375, 0.5},
		sizeType = Percent, size = {0.925, 0.9},
	},
}
```

This `TeButtonLayout`, `TeSpriteLayout`, etc are all supported in this PR.

**Review by sev- — COMMENTED** [review]
Incredible work. I started the review, but there are way too many files.

In general, the code is very clean.

**Comment by bluegr** [comment]
I started digging around on these games. Tetraedge was a company that ported games from PC to other platforms, using their own engine. This explains the differences in resource formats - they re-encoded the resources from the games they ported.

According to [MobyGames](https://www.mobygames.com/company/tetraedge-sa#), these are the games that they ported:

- Dracula III: The Path of the Dragon - Part 3 (2010)
- Dracula III: The Path of the Dragon - Part 2 (2010)
- Dracula III: The Path of the Dragon - Part 1 (2010)
- Return to Mysterious Island 2: Mina's Fate (2009)
- Safari Adventures: Africa (2009)
- Secrets of Da Vinci: The Forbidden Manuscript, The (2007)
- Syberia II (2007)
- Syberia (2006)
- Echo: Secrets of the Lost Cavern (2006)
- Return to Mysterious Island (2005)
- Egyptian Prophecy, The (2005)
- Colotiles (2002)

Their [Wikipedia page](https://fr.wikipedia.org/wiki/Tetraedge_Games) (French only) also contains these titles, among others:

- Atlantis III: The New World
- Return to Mysterious Island 2
- Egypt III

According to the Amerzone Wikipedia page, they have worked on the IOS port of the game:

https://en.wikipedia.org/wiki/Amerzone#Later_development

So, since they have reused code when porting these titles to other platforms, perhaps this engine could be extended to support them too :)

File: engines/tetraedge/game/character.cpp [hunk 1]
```diff
@@ -0,0 +1,1034 @@
...
+
+				int minoffset = 0;
+				for (int i = 0; i < 4; i++) {
+					if (framecounts[i] < framecounts[minoffset])
+						minoffset = i;
+				}
+
+				switch(minoffset) {
+				case 0:
+				  remainder = 29;
```
[the remark this answers is not in the document]
Comment (mduggan) [reply engines/tetraedge/game/character.cpp:993]: There was a bunch of places like this, I think I caught them all now.

File: engines/tetraedge/game/character.cpp [hunk 2]
```diff
@@ -0,0 +1,1034 @@
...
+	_curModelAnim.setDeleteFn(&TeModelAnimation::deleteLater);
+}
+
+Character::~Character() {
+	_model->setVisible(false);
+	_model->bonesUpdatedSignal().remove(this, &Character::onBonesUpdate);
+	deleteAnim();
+	Game *game = g_engine->getGame();
+	Common::Array<TeIntrusivePtr<TeModel>> &models = game->scene().models();
+	for (unsigned int i = 0; i < models.size(); i++) {
```
[the remark this answers is not in the document]
Comment (mduggan) [reply engines/tetraedge/game/character.cpp:80]: I had been inconsistent with this - switched them all to uint.

File: engines/tetraedge/detection.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "base/plugins.h"
+#include "common/config-manager.h"
```
[the remark this answers is not in the document]
Comment (mduggan) [reply engines/tetraedge/detection.cpp:23]: Yep, removed them.

File: engines/tetraedge/console.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
+/* ScummVM - Graphic Adventure Engine
```
[the remark this answers is not in the document]
Comment (mduggan) [reply engines/tetraedge/console.cpp:1]: Not needed - was left over from create_engine. Removed.

File: engines/tetraedge/detection_tables.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+const PlainGameDescriptor GAME_NAMES[] = {
+	{ "syberia", "Syberia" },
+	{ "amerzone", "Amerzone" },
+	{ "syberia2", "Syberia II" },
+	{ 0, 0 }
+};
+
+const ADGameDescription GAME_DESCRIPTIONS[] = {
+	{
+		"syberia",
```
[the remark this answers is not in the document]
Comment (mduggan) [reply engines/tetraedge/detection_tables.h:33]: Done.

File: engines/tetraedge/game/character.cpp [hunk 6]
```diff
@@ -0,0 +1,1034 @@
...
+#include "tetraedge/game/character_settings_xml_parser.h"
+#include "tetraedge/te/te_model_animation.h"
+#include "tetraedge/te/te_core.h"
+
+namespace Tetraedge {
+
+/*static*/ Common::HashMap<Common::String, Character::CharacterSettings> *Character::_globalCharacterSettings = nullptr;
+
+/*static*/ Common::Array<Character::AnimCacheElement> Character::_animCache;
+/*static*/ Common::HashMap<Common::String, TeIntrusivePtr<TeModelAnimation>> Character::_animCacheMap;
```
[the remark this answers is not in the document]
Comment (mduggan) [reply engines/tetraedge/game/character.cpp:41]: There were a few static objects - I've cleaned them all up now.

**Comment by mduggan** [comment]
> please switch to multi renderer model like in our engines. and current GL code as "classic GL" with up to 1.4 GL, best keep low as possible, also please avoid use GL begin/end

Done.  I added TGL support, and it switches based on configuration, but it's currently broken in software mode - the game is running but nothing is displayed.  I will probably need some help from someone who understands GL and TinyGL better than me, as I had to comment out a lot of env calls to get it compiling and running.

**Review by aquadran — APPROVED** [review]

**Pull request merged**